src/index.ts

```diff
@@ -419,6 +419,21 @@ class QueryPipeline extends EventEmitter {
         return obj;
       }, {});
 
+    let activePromotions = [] as ActivePromotion[];
+    let activePins = [] as string[];
+    if (jsonProto.activePromotions) {
+      activePromotions = jsonProto.activePromotions;
+      activePromotions.forEach((promotion) => {
+        if (promotion.activePins) {
+          promotion.activePins.forEach((pin) => {
+            if (activePins.indexOf(pin.key.value) === -1) {
```

Comment:
Looks like we're mostly using activePins to do indexOf lookups. I think you should use a Set: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set

Probably useful to read a bit about the performance characteristics of these two compared to one another.